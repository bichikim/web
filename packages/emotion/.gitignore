# modules
node_modules

# dest
/dist
