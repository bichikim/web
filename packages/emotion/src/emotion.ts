import {App, inject} from 'vue'
import createCache, {EmotionCache} from '@emotion/cache'

export const cacheSym = Symbol('cache')

interface VueEmotion {
  install: (app: App) => void
}

export const useCache = (): EmotionCache | null => {
  return inject(cacheSym, null)
}

export const createEmotion = (): VueEmotion => {
  return {
    install(app: App) {
      app.provide(cacheSym, createCache())
    },
  }
}

export default createEmotion
