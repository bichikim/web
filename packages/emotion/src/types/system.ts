import {AnyObject} from '@innovirus/utils'
import {CssFunctionReturnType, CSSProperties, CSSObject as _CSSObject} from '@styled-system/css'
import {Theme, styleFn} from 'styled-system'

export type PropsWithTheme<P, T extends Theme = Theme> = P & {
  theme?: T
}

export interface SystemFunc<P extends AnyObject, T extends Theme = Theme> {
  (props: PropsWithTheme<P, T>, ...args: any[]): _CSSObject | CssFunctionReturnType

  cache?: Record<string, any>
  config?: Record<string, any>
  propNames?: string[]
}

export type PureSystemFunc<P extends AnyObject, T extends Theme = Theme> = (props: PropsWithTheme<P, T>, ...args: any[]) => _CSSObject | CssFunctionReturnType;

export type CSSObject<P, T extends Theme = Theme> = _CSSObject | SystemFunc<P, T> | styleFn | PureSystemFunc<P, T>

export type StyledSystems<P, T extends Theme = Theme> = CSSObject<P, T>[]

export type System<P = AnyObject, T extends Theme = Theme> =
  ReadonlyArray<CSSObject<P, T> | ReadonlyArray<CSSObject<P, T>>>

export type FunctionCSSObject = (props: any) => any

export type CSSPropertiesWithMultiValues = {
  [K in keyof CSSProperties]: CSSProperties[K];
};
