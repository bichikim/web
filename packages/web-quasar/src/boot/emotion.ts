import {createEmotion} from '@winter-love/emotion'
import {BootCallback, BootFileParams} from '@quasar/app'
import {EmptyObject} from '@winter-love/utils'

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Express {
    interface Request {
      // eslint-disable-next-line
      __emotion_cache__: any
    }
  }
}

const emotionBoot: BootCallback<EmptyObject> = (context: BootFileParams<EmptyObject>) => {
  const {ssrContext, app} = context
  if (!ssrContext) {
    return
  }

  const emotion = createEmotion()

  app.use(emotion)

  const {req} = ssrContext

  req.__emotion_cache__ = emotion.cache
}

export default emotionBoot
