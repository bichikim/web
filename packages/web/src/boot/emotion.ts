import {createEmotion, EmotionPlugin} from '@winter-love/emotion'
import {BootCallback, BootFileParams} from '@quasar/app'
import {EmptyObject} from '@winter-love/utils'

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Express {
    interface Request {
      // eslint-disable-next-line
      __emotionCache__: any
    }

    interface Response {
      // eslint-disable-next-line
      __emotionCache__: any
    }
  }
}

let emotion: EmotionPlugin

export const _createEmotion = () => {
  if (emotion) {
    return emotion
  }
  emotion = createEmotion()
  return emotion
}

const emotionBoot: BootCallback<EmptyObject> = (context: BootFileParams<EmptyObject>) => {
  const {app} = context

  const emotion = _createEmotion()

  app.use(emotion)
}

export default emotionBoot
