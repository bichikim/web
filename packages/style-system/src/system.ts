import {AnyObject, PureObject} from '@winter-love/utils'
import {Properties} from 'csstype'
import {createParser} from './create-parser'
import {CSSObject} from '@emotion/css'
import {StyleParse} from './types'

export interface ConfigStyle<Theme extends AnyObject, Scale extends string | number | symbol = keyof Theme> {
  /** A fallback scale object for when there isn't one defined in the `theme` object. */
  defaultScale?: Scale
  /**
   * An array of multiple properties (e.g. `['marginLeft', 'marginRight']`) to which this style's value will be
   * assigned (overrides `property` when present).
   */
  properties?: Array<keyof Properties>
  /** The CSS property to use in the returned style object (overridden by `properties` if present). */
  property?: keyof Properties
  /** A string referencing a key in the `theme` object. */
  scale?: Scale
  /** A function to transform the raw value based on the scale. */
  transform?: (value: keyof Theme[Scale], scale?: Theme[Scale], props?: PureObject) => any
}
import {getScale} from './get-scale'

const getValue = (n: any, scale: any) => getScale(scale, n, n)

export const createStyleFunction = <Theme extends AnyObject, Scale extends string | number | symbol = keyof Theme>(options: ConfigStyle<Theme, Scale>) => {
  const {defaultScale, properties, property, scale, transform = getValue} = options
  const _properties = properties || [property]
  const sx = (value: any, scale: any, _props: any) => {
    const result: any = {}
    const n = transform(value, scale, _props)
    if (n === null) {
      return
    }
    _properties.forEach((prop: any) => {
      result[prop] = n
    })
    return result
  }
  sx.scale = scale
  sx.defaults = defaultScale
  return sx
}

const fillProperty = <Theme extends AnyObject = AnyObject>(option: ConfigStyle<Theme>, key: string): ConfigStyle<Theme> => {
  const {property, properties, ...rest} = option
  let _properties: any[]
  if (property) {
    _properties = [property]
  } else if (properties) {
    _properties = [...properties]
  } else {
    _properties = [key]
  }
  return {
    ...rest,
    properties: _properties,
  }
}

export interface SystemOptions<Theme extends PureObject = PureObject>{
  [customStyleName: string]: ConfigStyle<Theme> | boolean | ((value: any, scale: any, props: any, index: number) => CSSObject)
}

export const system = <Theme extends AnyObject>(options: SystemOptions<Theme> = {}): StyleParse => {
  const _options = {...options}
  const config = Object.keys(_options).reduce((config: Record<string, any>, key: string) => {
    const option: any = options[key]

    if (option === true) {
      config[key] = createStyleFunction({
        property: key as any,
        scale: key,
      })
      return config
    }

    if (typeof option === 'function') {
      config[key] = option
      return config
    }

    config[key] = createStyleFunction(fillProperty(option, key))
    return config
  }, {} as any)

  return createParser(config)
}
