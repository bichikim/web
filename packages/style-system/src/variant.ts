import {StyleParse} from 'src/types'
import {CSSObject} from '@emotion/css'
import {system} from './system'

export interface VariantOptions {
  prop: string
  scale?: string
  variants?: Record<string, CSSObject>
}

const deepParse = (style: Record<string, any>, parse, theme) => {
  const {deeps, left} = Object.keys(style).reduce((result, key) => {
    const value = style[key]
    if (typeof value === 'object' && !Array.isArray(value)) {
      result.deeps[key] = parse({...value, theme})
      return result
    }
    result.left[key] = value
    return result
  }, {deeps: {}, left: {}})

  return {
    ...parse({...left, theme}),
    ...deeps,
  }
}

export const createVariant = (parse: StyleParse) => (options: VariantOptions) => {
  const {prop, scale, variants} = options
  return system({
    [prop]: Object.assign((value: any, scale: any, props: any) => {
      const {theme} = props
      const style = scale?.[value]
      return typeof style === 'object' ? deepParse(style, parse, theme) : {}
    }, {
      defaults: variants,
      scale,
    }),
  })
}
