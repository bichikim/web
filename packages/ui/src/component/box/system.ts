import {
  allSystem,
  AllSystemProps,
  ResponsiveValue,
  ShadowProps,
  show,
  ShowProps,
  Theme,
} from 'src/systems'
import {ASProps, System} from 'src/types'
import {variantComplex} from 'src/utils'

export interface textSetVariantProps {
  textSet?: ResponsiveValue<string>
}

export interface RefProps {
  ref?: any
}

export type Props =
  & ASProps
  & RefProps
  & ShadowProps
  & ShowProps
  & textSetVariantProps
  & AllSystemProps

export const boxSystem: System<Props, Theme> = [
  {
    boxSizing: 'border-box',
    display: 'block',
    width: '100%',
    color: 'inherit',
    textDecoration: 'inherit',
    height: '100%',
    padding: '0',
    border: 0,
    margin: 0,
    backgroundColor: 'transparent',
    outline: 'none !important',
    fontSize: '1rem',
    overflow: 'hidden',
    '&:visited': {

    },
  },
  allSystem,
  show,
  variantComplex({
    prop: 'textSet',
    scale: 'typography',
  }),
]
