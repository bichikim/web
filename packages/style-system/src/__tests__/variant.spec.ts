/**
 * @jest-environment jsdom
 */

import {createVariant} from '../variant'
import {system} from '../system'
import {createEmotion, Theme} from '@winter-love/emotion'
import {mount} from '@vue/test-utils'
import {h} from 'vue-demi'

const theme: Theme = {
  breakpoints: ['500px', '750px', '1200px'],
  space: {
    lg: '15px',
    md: '10px',
    sm: '5px',
  },
}

const emotion = createEmotion({theme})

const themeVariants = {
  ...theme,
  variants: {
    deep: {
      '&>div': {
        '.foo': {
          padding: 'sm',
        },
        margin: 'sm',
      },
      margin: 'lg',
      padding: 'md',
    },
    lg: {
      margin: 'lg',
      padding: 'md',
    },
    md: {
      margin: 'lg',
      padding: 'sm',
    },
    sm: {
      margin: 'sm',
      padding: 'sm',
    },
  },
}

const systems = system({
  margin: {
    scale: 'space',
  },
  padding: {
    scale: 'space',
  },
})

const setup = () => {
  return createVariant(systems)
}

describe('variant', () => {
  it('should return styles', () => {
    const variant = setup()

    const variantParse = variant({
      prop: 'variant',
      scale: 'variants',
      variants: {
        lg: {
          margin: 'lg',
          padding: 'lg',
        },
      },
    })

    const Component = emotion.styled('div', {
      stylePortal: 'css',
    })(variantParse, () => {
      return {}
    })

    const wrapper = mount(Component, {
      props: {css: {foo: 'bar', variant: 'lg'}},
    })

    expect(wrapper.element).toHaveStyle({
      margin: '15px',
      padding: '15px',
    })
    {
      const result = variantParse({
        theme: themeVariants,
        variant: 'md',
      })

      expect(result).toEqual({
        margin: '15px',
        padding: '5px',
      })
    }
    {
      const result = variantParse({
        theme: themeVariants,
        variant: 'sm',
      })

      expect(result).toEqual({
        margin: '5px',
        padding: '5px',
      })
    }
    {
      const result = variantParse({
        theme: themeVariants,
        variant: 'foo',
      })

      expect(result).toEqual({})
    }
    {
      const result = variantParse({
        theme,
        variant: 'lg',
      })

      expect(result).toEqual({
        margin: '15px',
        padding: '15px',
      })
    }
  })
  it('should not return styles', () => {
    const variant = setup()

    const variantParse = variant({
      prop: 'variant',
      scale: 'variants',
    })

    {
      const result = variantParse({
        theme,
        variant: 'lg',
      })

      expect(result).toEqual({})
    }
  })
  it.skip('should not return styles', () => {
    const variant = setup()

    const variantParse = variant({
      prop: 'variant',
      scale: 'variants',
    })

    const Component = emotion.styled('div', {
      stylePortal: 'css',
    })(
      variantParse,
      systems,
    )

    const Component2 = (props) => {
      return h(Component, props, () => [
        h('div', {class: 'bar'}, [
          h('div', {class: 'foo'}),
        ]),
      ])
    }

    const wrapper = mount(Component2, {
      props: {css: {margin: 'sm', variant: 'deep'}, theme: themeVariants},
    })

    expect(wrapper.element).toMatchSnapshot()

    expect(wrapper.element).toHaveStyle({
      margin: '5px',
      padding: '10px',
    })

    expect(wrapper.get('.bar').element).toHaveStyle({
      margin: '5px',
    })

    expect(wrapper.get('.foo').element).toHaveStyle({
      padding: '5px',
    })
  })
})
