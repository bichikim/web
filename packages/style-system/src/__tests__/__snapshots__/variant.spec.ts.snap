// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variant should not return styles 1`] = `
.emotion-0 {
  margin: 15px;
  padding: 10px;
  margin: 5px;
}

.emotion-0>div {
  margin: 5px;
}

.emotion-0>div .foo {
  padding: 5px;
}

<div
  class=" emotion-0"
>
  <div
    class="bar"
  >
    <div
      class="foo"
    />
  </div>
</div>
`;
